components/tidb_query_expr/src/impl_like.rs

```diff
@@ -66,6 +70,87 @@ pub fn like<C: Collator>(target: BytesRef, pattern: BytesRef, escape: &i64) -> R
     Ok(Some(true as i64))
 }
 
+#[rpn_fn(capture = [metadata], metadata_mapper = init_regexp_utf8_data)]
+#[inline]
+pub fn regexp_utf8(
+    metadata: &Option<Regex>,
+    target: BytesRef,
+    pattern: BytesRef,
+) -> Result<Option<i64>> {
+    let target = match String::from_utf8(target.to_vec()) {
+        Ok(target) => target,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+
+    match metadata {
+        Some(regex) => Ok(Some(regex.is_match(target.as_ref()) as i64)),
+        None => {
+            let pattern = match String::from_utf8(pattern.to_vec()) {
+                Ok(pattern) => pattern,
+                Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+            };
+            let pattern = format!("(?i){}", pattern);
+
+            match BytesRegex::new(&pattern) {
+                Ok(bytes_regex) => Ok(Some(bytes_regex.is_match(target.as_ref()) as i64)),
+                Err(err) => Err(box_err!("invalid regex pattern: {:?}", err)),
+            }
+        }
+    }
+}
+
+#[rpn_fn(capture = [metadata], metadata_mapper = init_regexp_data)]
+#[inline]
+pub fn regexp(
+    metadata: &Option<BytesRegex>,
+    target: BytesRef,
+    pattern: BytesRef,
+) -> Result<Option<i64>> {
+    match metadata {
+        Some(regex) => Ok(Some(regex.is_match(target.as_ref()) as i64)),
+        None => {
+            let pattern = match String::from_utf8(pattern.to_vec()) {
+                Ok(pattern) => pattern,
+                Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+            };
+            match BytesRegex::new(&pattern) {
+                Ok(bytes_regex) => Ok(Some(bytes_regex.is_match(target.as_ref()) as i64)),
+                Err(err) => Err(box_err!("invalid regex pattern: {:?}", err)),
+            }
+        }
+    }
+}
+
+fn init_regexp_utf8_data(expr: &mut Expr) -> Result<Option<Regex>> {
+    let children = expr.mut_children();
+    let n = children.len();
+    assert!(n > 1);
+
+    let tree_node = &mut children[1];
+    match tree_node.get_tp() {
+        ExprType::ScalarFunc | ExprType::ColumnRef | ExprType::Null => Ok(None),
+        _ => match Regex::new(String::from_utf8(tree_node.take_val()).unwrap().as_ref()) {
+            Ok(regex) => Ok(Some(regex)),
+            Err(_) => Ok(None),
+        },
+    }
+}
+
+fn init_regexp_data(expr: &mut Expr) -> Result<Option<BytesRegex>> {
+    let children = expr.mut_children();
+    let n = children.len();
+    assert!(n > 1);
+
+    let tree_node = &mut children[1];
+    match tree_node.get_tp() {
+        ExprType::ScalarFunc | ExprType::ColumnRef | ExprType::Null => Ok(None),
+        _ => match BytesRegex::new(String::from_utf8(tree_node.take_val()).unwrap().as_ref()) {
+            Ok(regex) => Ok(Some(regex)),
+            Err(_) => Ok(None),
+        },
+    }
+}
```

Comment:
```suggestion
    match tree_node.get_tp() {
         ExprType::Bytes => {
              let pattern = match String::from_utf8(tree_node.take_val()) {
                   Ok(pattern) => pattern,
                   Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
              };
              match BytesRegex::new(&pattern) {
                Ok(bytes_regex) => Ok(Some(bytes_regex.is_match(target.as_ref()) as i64)),
                Err(err) => Err(box_err!("invalid regex pattern: {:?}", err)),
            }
         }
         _ => Ok(None)
    }
```

Reply:
done

Reply:
BTW, in my comprehension `RpnFnScalarEvaluator::evaluate_raw` just sets field type and tp of each param, and does not set val of param; which will make test fail (`init_regexp_data` / `init_regexp_utf8_data` just read val part of the param as the `pattern`, and this will get a null value). Is it reasonable? Or the code is misunderstood?

https://github.com/tikv/tikv/blob/master/components/tidb_query_expr/src/types/test_util.rs#L91

Reply:
But seems your tests are passing. What's wrong with the `evaluate_raw`?

Reply:
`evaluate_raw` does not set val of param, which makes the Regex will have a null value to `new`; while I'm not sure whether set value for `evaluate_raw` will make other side effect. Thus in the case `RpnExpression` is just generated manually.